Highbar.Types/Either.cs

```diff
@@ -0,0 +1,365 @@
+    /// </summary>
+    public abstract V Fold<V>(Func<L, V> leftMapper, Func<R, V> rightMapper);
+
+    /// <summary>
+    /// Gets the left value regardless of <c>Left</c> or <c>Right</c> status.
+    /// </summary>
+    public abstract L GetLeft();
+
+    /// <summary>
+    /// Gets the right value regardless of <c>Left</c> or <c>Right</c> status.
+    /// </summary>
+    public abstract R GetRight();
+
+    /// <summary>
+    /// Returns the value if the <c>instance</c> is a <c>Right</c>, otherwise the <c>otherValue</c>.
+    /// </summary>
+    public abstract R GetOrElse(R otherValue);
+
+    /// <summary>
+    /// Returns the value if the <c>instance</c> is a <c>Right</c>, otherwise the left-applied value of the <c>mapper</c>.
+    /// </summary>
+    public abstract R GetOrElseGet(Func<L, R> mapper);
+
+    /// <summary>
+    /// Returns the value if the <c>instance</c> is a <c>Right</c>, otherwise the value of the <c>supplier</c>.
+    /// </summary>
+    public abstract R GetOrElseGet(Func<R> supplier);
+
+    /// <summary>
+    /// Returns the value if the <c>instance</c> is a <c>Right</c>, otherwise throws an <c>Exception</c> mapped from the <c>Left</c> value.
+    /// </summary>
+    public abstract R GetOrElseThrow(Func<L, Exception> mapper);
+
+    /// <summary>
+    /// Returns the value if the <c>instance</c> is a <c>Right</c>, otherwise throws the supplied <c>Exception</c>.
+    /// </summary>
+    public abstract R GetOrElseThrow(Func<Exception> supplier);
+
+    /// <summary>
+    /// Executes the <c>consumer</c> if the <c>instance</c> is a <c>Left</c>.
+    /// </summary>
+    public abstract Either<L, R> IfLeft(Action<L> consumer);
+
+    /// <summary>
+    /// Executes the <c>consumer</c> if the <c>instance</c> is a <c>Right</c>.
+    /// </summary>
+    public abstract Either<L, R> IfRight(Action<R> consumer);
+
+    /// <summary>
+    /// Determines whether or not the <c>instance</c> is a <c>Left</c>.
+    /// </summary>
+    public abstract bool IsLeft
+    {
+      get;
+    }
+
+    /// <summary>
+    /// Determines whether or not the <c>instance</c> is a <c>Right</c>.
+    /// </summary>
+    public abstract bool IsRight
+    {
+      get;
+    }
+
+    /// <summary>
+    /// See Either#first(Func) first.
+    /// </summary>
+    public Either<ML, R> LeftMap<ML>(Func<L, ML> mapper)
+    {
+      return First(mapper);
+    }
+
+    /// <summary>
+    /// Functor <c>map</c>.  Takes a <c>Func</c> to map the biased right <c>value</c>.
+    /// </summary>
+    public abstract Either<L, MR> Map<MR>(Func<R, MR> mapper);
+
+    /// <summary>
+    /// Recover from a <c>Left</c> into a possible <c>Right</c>.
+    /// </summary>
+    public abstract Either<L, R> Recover(R value);
+
+    /// <summary>
+    /// Recover from a <c>Left</c> into a <c>Right</c> with a value produced by a <c>Func</c>.
+    /// </summary>
+    public abstract Either<L, R> Recover<A>(Func<A, R> other);
```

Comment:
It matches what's in the Java library.  It's a function where we don't really care what the input argument type is.